### ISSUE TITLE
Installing nvcc_linux-64 will drop ssh connection

### ISSUE THREAD

**Issue by pearu** [conda-forge/nvcc-feedstock#22]
Issue:
```
conda create -n test-nvvc -y
conda activate test-nvvc
export CUDA_HOME=/usr/local/cuda
export PATH=$CUDA_HOME/bin:$PATH
which cuda-gdb  # -> /usr/local/cuda/bin/cuda-gdb
conda install -c conda-forge nvcc_linux-64
```
will fail with
```
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: /home/pearu/miniconda3/envs/test-nvvc

  added / updated specs:
    - nvcc_linux-64


The following NEW packages will be INSTALLED:

  nvcc_linux-64      conda-forge/linux-64::nvcc_linux-64-10.1-h51cf6c1_3


Proceed ([y]/n)? 

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
--print-sysroot: command not found
Directory  doesn't exist
Connection to <ssh host> closed.
```

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info

     active environment : base
    active env location : /home/pearu/miniconda3
            shell level : 1
       user config file : /home/pearu/.condarc
 populated config files : /home/pearu/.condarc
          conda version : 4.7.12
    conda-build version : 3.18.9
         python version : 3.7.3.final.0
       virtual packages : __cuda=10.1
       base environment : /home/pearu/miniconda3  (writable)
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /home/pearu/miniconda3/pkgs
                          /home/pearu/.conda/pkgs
       envs directories : /home/pearu/miniconda3/envs
                          /home/pearu/.conda/envs
               platform : linux-64
             user-agent : conda/4.7.12 requests/2.22.0 CPython/3.7.3 Linux/4.15.0-62-generic ubuntu/18.04.3 glibc/2.27
                UID:GID : 1001:1001
             netrc file : None
           offline mode : False
```
</details>

**Comment by jakirkham** [comment]
Yeah this is because the package expects the conda C compiler to be available to symlink `libcuda.so` into it's sysroot. If you run `conda install -c conda-forge compilers` beforehand, that *should* work. Though please let us know either way.

**Comment by pearu** [comment]
Neither
```
conda install -c conda-forge compilers
conda install -c conda-forge nvcc_linux-64
```
nor
```
conda install -c conda-forge compilers nvcc_linux-64
```
helped - same failure.

However, after re-login and
```
export CUDA_HOME=/usr/local/cuda
export PATH=$CUDA_HOME/bin:$PATH
conda activate test-nvvc
```
the conda env seems to initialize ok.

**Comment by jakirkham** [comment]
What if you run `conda activate ...` between the two installs?

**Comment by pearu** [comment]
No change when doing deactivate and re-activate.

Since the activation scripts are sourced, the scripts should not execute `exit 1` statements that will lead to exiting the current shell. Instead, the nvcc activate script ought to use bash functions for implementing its functionality and error handing.

Although, this does not explain why `CC` is not defined inside the nvcc activate script when installing while it is defined when activating the environment. I'll continue debugging this tomorrow.

**Comment by jakirkham** [comment]
Have been bitten by similar `ssh` termination issues from the current handling strategy. Would be happy to review a PR if you want to try out some alternative error handling strategies.

Agree that's puzzling. Help debugging this would be welcome. 🙂

**Comment by jakirkham** [comment]
One guess would be that Conda is activating these in the opposite order for some reason. I'm not entirely sure why that is.

There was discussion about sorting activation scripts in dependency order ( https://github.com/conda/conda/issues/6338 ), but that has not happened AFAIK and wouldn't help us here (as the C compiler is not a dependency). Conda does sort them in alphabetic order ( https://github.com/conda/conda/pull/7176 ) to avoid a totally random activation order. Though I would think the the compilers' activation scripts (namely `activate-binutils_linux-64.sh`, `activate-gfortran_linux-64.sh`, `activate-gcc_linux-64.sh`, `activate-gxx_linux-64.sh`) would come before our activation script (namely `nvcc_linux-64_activate.sh`). Probably missing something. 🤔

**Comment by jakirkham** [comment]
Added PR ( https://github.com/conda-forge/nvcc-feedstock/pull/24 ) to avoid exiting from the current session. This doesn't solve the underlying activation problem, but should make it easier to debug at least. Not to mention be better behaved.

**Comment by pearu** [comment]
I can confirm that the `activate*.sh` scripts are run before `nvcc*_activate.sh`.

If `CC` would be the only required variable for nvcc activate the script, then one workaround would be to use:
```
CONDA_ENV_SYSROOT="\$PREFIX/\$HOST/sysroot"
```
I cannot think of a case where `"\$(\${CC} --print-sysroot)"` would give a different result (provided that `CC` is from conda `compilers` package).

**Comment by jakirkham** [comment]
It might be worth hacking the activate scripts in your environment to print out debug info like `env`.

**Comment by pearu** [comment]
So, when executing `conda install nvcc_linux-64` and looking at the `env` output inside the nvcc activate script, I see that `CUDA_HOME` environment variable is reset to empty value which will lead to a message: `Directory  doesn't exist`.

**Comment by jakirkham** [comment]
Yeah I’ve been seeing that too. Is this on a box or are using one of conda-forge’s Docker images?

**Comment by pearu** [comment]
It's a local Ubuntu 18.04 box.

**Comment by pearu** [comment]
I think I found what is the problem: when running `conda install nvcc_linux-64`, at some point `nvcc` deactivate script is run *before* the activate script. The deactivate script will set `CUDA_HOME="${CUDA_HOME_CONDA_NVCC_BACKUP}"` where `CUDA_HOME_CONDA_NVCC_BACKUP` is empty.

**Comment by pearu** [comment]
Using
```
if [ -z "${CUDA_HOME_CONDA_NVCC_BACKUP+x}" ]; then
  export CUDA_HOME="${CUDA_HOME_CONDA_NVCC_BACKUP}"
fi
```
to nvcc deactivate script appears to fix the issue. Similar if-blocks should be added for other `*CONDA_NVCC_BACKUP` variable cases.

**Comment by pearu** [comment]
In addition, the following code in nvcc deactivate 
```
# Remove $(libcuda.so) shared object stub from the compiler sysroot.
CONDA_ENV_SYSROOT="$(${CC} --print-sysroot)"
rm -f "${CONDA_ENV_SYSROOT}/lib/libcuda.so"
```
is dangerous when one has defined `CC=/path/to/system/gcc` and has not installed conda `compilers` package. The result is:
```
ln: failed to create symbolic link '//lib/libcuda.so': Permission denied
```
So, I thing using `CONDA_ENV_SYSROOT="\$PREFIX/\$HOST/sysroot"` would be a saver solution in many ways (does not use CC, avoids an attempt to remove `libcuda.so`).

**Comment by jakirkham** [comment]
I've raised issue ( https://github.com/conda-forge/nvcc-feedstock/issues/25 ) to discuss adding the conda compilers as explicit dependencies.

**Comment by pearu** [comment]
Resolved by #24 and #27.
